user-service/src/main/java/org/java/mentorship/user/service/UserService.java

```diff
@@ -58,12 +62,94 @@ public boolean verifyUserHash(Integer id, String password) {
     }
 
     public boolean verifyUserUsingToken(Integer id, String token) {
-        Optional<UserEntity> user = getUserById(id);
-        if (user.isEmpty()) throw new UserNotFoundException();
-        if (!user.get().getVerificationToken().equals(token)) {
+        UserEntity user = getUserById(id).orElseThrow(UserNotFoundException::new);
+        if (!user.getVerificationToken().equals(token)) {
             return false;
         }
 
-        return mapper.setUserVerifiedStatus(id, true);
+        if (user.getLastSentVerificationNotification() != null)
+            if (OffsetDateTime.now(ZoneOffset.UTC).isAfter(user.getLastSentVerificationNotification().plusMinutes(10)))
+                return false;
+
+        user.setVerifiedAt(OffsetDateTime.now(ZoneOffset.UTC));
+        mapper.update(user);
+
+        return true;
+    }
+
+    public Boolean resendVerificationToken(Integer userId) {
+        UserEntity user = this.getUserById(userId).orElseThrow(UserNotFoundException::new);
+
+        if (user.getLastSentVerificationNotification() != null)
+            if (!OffsetDateTime.now(ZoneOffset.UTC).isAfter(user.getLastSentVerificationNotification().plusMinutes(10)))
+                return false;
+
+        user.setVerificationToken(UUID.randomUUID().toString());
+        user.setLastSentVerificationNotification(OffsetDateTime.now(ZoneOffset.UTC));
+
+        notificationFeignClient.postNotification(new Notification(
+                user.getId(), user.getEmail(),
+                Collections.singletonList(NotificationChannel.EMAIL), NotificationType.VERIFICATION,
+                Map.of(
+                        "firstName", user.getFirstName(),
+                        "lastName", user.getLastName(),
+                        "verificationToken", user.getVerificationToken(),
+                        "requestedAt", OffsetDateTime.now(ZoneOffset.UTC)
+                )
+        ));
+
+        mapper.update(user);
+
+        return true;
+    }
+
+    public Boolean changePassword(Integer userId, String newPassword) {
+        UserEntity user = this.getUserById(userId).orElseThrow(UserNotFoundException::new);
+
+        user.setHashedPassword(getMd5(newPassword));
+        user.setLastChangedPassword(OffsetDateTime.now(ZoneOffset.UTC));
+        user.setPasswordChangeToken(null);
+
+        mapper.update(user);
+
+        return true;
+    }
+
+    public Boolean changePasswordWithToken(Integer userId, String newPassword, String token) {
+        UserEntity user = this.getUserById(userId).orElseThrow(UserNotFoundException::new);
+
+        if (user.getLastSentPasswordChangeToken() != null)
+            if (OffsetDateTime.now(ZoneOffset.UTC).isAfter(user.getLastSentPasswordChangeToken().plusMinutes(1)))
+                return false;
+
+        if (!Objects.equals(user.getPasswordChangeToken(), token)) return false;
+
+        return changePassword(userId, newPassword);
+    }
+
+    public Boolean requestChangePasswordToken(Integer userId) {
```

Comment:
I think we could extract everything related to tokens (verification, password recovery) in a separate service (TokenService) and make that service call the UserService